frontend/lib/resolution/scope-queries.cpp

```diff
@@ -1623,6 +1623,97 @@ lookupNameInScopeWithSet(Context* context,
   return vec;
 }
 
+static std::map<UniqueString, BorrowedIdsWithName> const&
+getSymbolsAvailableInScopeQuery(Context* context,
+                                const Scope* scope,
+                                const VisibilitySymbols* inVisibilitySymbols) {
+  QUERY_BEGIN(getSymbolsAvailableInScopeQuery, context, scope, inVisibilitySymbols);
+
+  std::map<UniqueString, BorrowedIdsWithName> toReturn;
+
+  auto allowedByVisibility = [inVisibilitySymbols](UniqueString name, UniqueString& renameTo) {
+    renameTo = name;
+
+    if (!inVisibilitySymbols) return true;
+    auto kind = inVisibilitySymbols->kind();
+
+    if (kind == VisibilitySymbols::ALL_CONTENTS) {
+      return true;
+    }
+
+    if (kind == VisibilitySymbols::ONLY_CONTENTS ||
+        kind == VisibilitySymbols::CONTENTS_EXCEPT) {
+      auto& namePairs = inVisibilitySymbols->names();
+
+      bool anyMatches = false;
+      for (auto& namePair : namePairs) {
+        if (namePair.first != name) continue;
+
+        anyMatches = true;
+        renameTo = namePair.second;
+        break;
+      }
+
+      return kind == VisibilitySymbols::CONTENTS_EXCEPT ? !anyMatches : anyMatches;
+    }
+
+    return false;
+  };
+
+  for (auto& decl : scope->declared()) {
+    UniqueString renameTo;
+    if (!allowedByVisibility(decl.first, renameTo)) continue;
+
+    auto flagSet = 0;
```

Comment:
should `flagSet` be initialized to `IdAndFlags::FlagSet::empty()` as well?

Reply:
No, they are different types. It shouldn't be called `flagSet`, though. In short: we have a single 'inclusive' filter (i.e., allowlist), and many 'exclusive' filters (i.e., blocklists). The reason for this is that we can't represent all necessary search patterns using a single bitfield.